crates/nu-parser/src/deparse.rs

```diff
@@ -1,14 +1,24 @@
-pub fn escape_quote_string(input: &str) -> String {
-    let mut output = String::with_capacity(input.len() + 2);
-    output.push('"');
-
+fn parse_input(input: &str) -> String {
+    let mut output = String::with_capacity(input.len());
     for c in input.chars() {
         if c == '"' || c == '\\' {
             output.push('\\');
         }
         output.push(c);
     }
-
-    output.push('"');
     output
 }
+
+pub fn escape_quote_string(input: &str) -> String {
+    if input.starts_with('-') {
+        parse_input(input)
```

Comment:
If we run `nu script.nu --foo='bar baz' qux`, `input` will be `"--foo=bar baz"` and `"qux"`, and the results will be the same. This is incorrect because the results will then be concatenated into `"main --foo=bar baz qux"` and evaluated as a nu command. Here, `"--foo=bar baz"` is split into two separate arguments `"--foo=bar"` and `"baz"` to the main function in script.nu.

The input `"--foo=bar baz"` should be escaped as `"--foo=\"bar baz\""` in this case.

Reply:
Ah, I see. I'll keep working at it.

Reply:
Found another corner case with the proposed approach. That is, values start with `-`cannot be passed as positional arguments.

```
bash-3.2$ cat script.nu
def main [--foo: string, arg: string] {
  echo $foo
  echo $arg
}

bash-3.2$ target/debug/nu script.nu --foo=bar -this-is-arg-
Error: nu::parser::unknown_flag (https://docs.rs/nu-parser/0.61.1/nu_parser/enum.ParseError.html#variant.UnknownFlag)

  × The `main` command doesn't have flag `-t`.
   ╭─[<commandline>:1:1]
 1 │ main --foo=bar -this-is-arg-
   ·                 ┬
   ·                 ╰── unknown flag
   ╰────
  help: use main --help for a list of flags

bash-3.2$ target/debug/nu script.nu --foo=bar -- -this-is-arg-
Error: nu::parser::unknown_flag (https://docs.rs/nu-parser/0.61.1/nu_parser/enum.ParseError.html#variant.UnknownFlag)

  × The `main` command doesn't have flag ``.
   ╭─[<commandline>:1:1]
 1 │ main --foo=bar -- -this-is-arg-
   ·                ─┬
   ·                 ╰── unknown flag
   ╰────
  help: use main --help for a list of flags
```

Reply:
```
bash-3.2$ cat script.nu
def main [--foo: string, arg: string] {
  echo $foo
  echo $arg
}

bash-3.2$ target/debug/nu script.nu --foo=bar -this-is-arg-
```

is working for me now. Is it working for you, @uasi?

![image](https://user-images.githubusercontent.com/88221256/167053672-3f612993-c888-498a-b933-48cd028f5b68.png)

Reply:
@merelymyself It works, but still fails with `--this-is-arg--`.

```
bash-3.2$ target/debug/nu script.nu --foo=bar --this-is-arg--
Error: nu::parser::unknown_flag (https://docs.rs/nu-parser/0.62.0/nu_parser/enum.ParseError.html#variant.UnknownFlag)

  × The `main` command doesn't have flag `this-is-arg--`.
   ╭─[<commandline>:1:1]
 1 │ main --foo="bar" --this-is-arg--
   ·                  ───────┬───────
   ·                         ╰── unknown flag
   ╰────
  help: use main --help for a list of flags
```

I doubt older versions of nu can handle it, so I think we can go ahead with this PR and fix it later.

Reply:
To be precise, nu _functions_ can take values that start with `--` as arguments if they are quoted, but nu _scripts_ cannot.

```
〉def f [arg: string] { echo $arg }
〉f --this-is-arg--
Error: nu::parser::unknown_flag (https://docs.rs/nu-parser/0.62.0/nu_parser/enum.ParseError.html#variant.UnknownFlag)

  × The `f` command doesn't have flag `this-is-arg--`.
   ╭─[entry #2:1:1]
 1 │ f --this-is-arg--
   ·   ───────┬───────
   ·          ╰── unknown flag
   ╰────
  help: use f --help for a list of flags

〉f "--this-is-arg--"
--this-is-arg--
```

```
〉target/debug/nu script.nu --foo=bar "--this-is-arg--"                                                                               
  × The `main` command doesn't have flag `this-is-arg--`.
   ╭─[<commandline>:1:1]
 1 │ main --foo="bar" --this-is-arg--
   ·                  ───────┬───────
   ·                         ╰── unknown flag
   ╰────
  help: use main --help for a list of flags
```

Reply:
That is an issue. I agree that that's most likely out of the scope of this PR, because I don't think older versions of nu handled it appropriately either. Still, I'll look more into it!